core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1936,7 +1984,311 @@ protected Scope visitUpdate(Update update, Optional<Scope> scope)
+                            throw semanticException(TYPE_MISMATCH, predicate, "WHERE clause must evaluate to a boolean: actual type %s", predicateType);
+                        }
+                        // coerce the predicate to boolean
+                        analysis.addCoercion(predicate, BOOLEAN, false);
+                    }
+                }
+            }
+
+            // Add any necessary casts of column expressions to column type, and throw an exception
+            // if the expression type cannot be cast to the column type
+            for (int caseIndex = 0; caseIndex < merge.getMergeCases().size(); caseIndex++) {
+                MergeCase mergeCase = merge.getMergeCases().get(caseIndex);
+                if (mergeCase instanceof MergeDelete) {
+                    continue;
+                }
+                ImmutableList.Builder<Type> setColumnTypesBuilder = ImmutableList.builder();
+                ImmutableList.Builder<Type> setExpressionTypesBuilder = ImmutableList.builder();
+                List<String> columnList = mergeCaseColumnsLists.get(caseIndex);
+                allColumnTypes.forEach((name, type) -> {
+                    int index = columnList.indexOf(name);
+                    if (index >= 0) {
+                        setColumnTypesBuilder.add(type);
+                        Expression expression = requireNonNull(mergeCase.getSetExpressions().get(index), "merge set expression is null");
+                        Type expressionType = analysis.getType(expression);
+                        setExpressionTypesBuilder.add(expressionType);
+                    }
+                });
+                List<Type> setColumnTypes = setColumnTypesBuilder.build();
+                List<Type> setExpressionTypes = setExpressionTypesBuilder.build();
+                if (!typesMatchForInsert(setColumnTypes, setExpressionTypes)) {
+                    throw semanticException(TYPE_MISMATCH,
+                            mergeCase,
+                            "MERGE table column types don't match for MERGE case %s, SET expressions: Table: [%s], Expressions: [%s]",
+                            caseIndex,
+                            Joiner.on(", ").join(setColumnTypes),
+                            Joiner.on(", ").join(setExpressionTypes));
+                }
+                for (int index = 0; index < setColumnTypes.size(); index++) {
+                    Expression expression = mergeCase.getSetExpressions().get(index);
+                    Type targetType = setColumnTypes.get(index);
+                    Type expressionType = setExpressionTypes.get(index);
+                    if (!targetType.equals(expressionType)) {
+                        analysis.addCoercion(expression, targetType, typeCoercion.isTypeOnlyCoercion(expressionType, targetType));
+                    }
+                }
+            }
+
+            // The final version of MergeAnalysis, with the finalQuery
```

Comment:
The `finalQuery` should not be created in the Analyzer. This belongs to the Planner. cc @martint

Reply:
Let's get @martint's take on it.